From 4b5718dd11e4caa391f793666cf523a4c7d5173e Mon Sep 17 00:00:00 2001
From: Yu Chien Peter Lin <peterlin@andestech.com>
Date: Mon, 21 Aug 2023 16:38:36 +0800
Subject: [PATCH 048/241] target/riscv: Add initial support for AX45MPV

This patch adds support for the AX45MPV, which implements
the vector extension (spec v1.0) with 1024 VLEN and 64 ELEN,
along with support for the bit manipulation extension (Zbabcs).

Signed-off-by: Yu Chien Peter Lin <peterlin@andestech.com>
---
 target/riscv/cpu-qom.h |  3 ++-
 target/riscv/cpu.c     | 25 +++++++++++++++++++++++++
 2 files changed, 27 insertions(+), 1 deletion(-)

diff --git a/target/riscv/cpu-qom.h b/target/riscv/cpu-qom.h
index 58a9fab41d..a539392a62 100644
--- a/target/riscv/cpu-qom.h
+++ b/target/riscv/cpu-qom.h
@@ -33,13 +33,14 @@
 #define TYPE_RISCV_CPU_ANDES_AX25       RISCV_CPU_TYPE_NAME("andes-ax25")
 #define TYPE_RISCV_CPU_ANDES_AX27       RISCV_CPU_TYPE_NAME("andes-ax27")
 #define TYPE_RISCV_CPU_ANDES_AX45       RISCV_CPU_TYPE_NAME("andes-ax45")
+#define TYPE_RISCV_CPU_ANDES_AX45MPV    RISCV_CPU_TYPE_NAME("andes-ax45mpv")
 #define TYPE_RISCV_CPU_ANDES_AX65       RISCV_CPU_TYPE_NAME("andes-ax65")
 #define TYPE_RISCV_CPU_ANDES_N25        RISCV_CPU_TYPE_NAME("andes-n25")
 #define TYPE_RISCV_CPU_ANDES_N45        RISCV_CPU_TYPE_NAME("andes-n45")
 #define TYPE_RISCV_CPU_ANDES_NX25       RISCV_CPU_TYPE_NAME("andes-nx25")
 #define TYPE_RISCV_CPU_ANDES_NX27V      RISCV_CPU_TYPE_NAME("andes-nx27v")
 #define TYPE_RISCV_CPU_ANDES_NX45       RISCV_CPU_TYPE_NAME("andes-nx45")
-#define TYPE_RISCV_CPU_ANDES_NX45V_META RISCV_CPU_TYPE_NAME("andes-nx45v-meta")
+#define TYPE_RISCV_CPU_ANDES_NX45V      RISCV_CPU_TYPE_NAME("andes-nx45v")
 #define TYPE_RISCV_CPU_ANY              RISCV_CPU_TYPE_NAME("any")
 #define TYPE_RISCV_CPU_MAX              RISCV_CPU_TYPE_NAME("max")
 #define TYPE_RISCV_CPU_BASE32           RISCV_CPU_TYPE_NAME("rv32")
diff --git a/target/riscv/cpu.c b/target/riscv/cpu.c
index c0f77f057d..2a9ea80e52 100644
--- a/target/riscv/cpu.c
+++ b/target/riscv/cpu.c
@@ -632,6 +632,30 @@ static void rv64_andes_ax45_cpu_init(Object *obj)
     cfg->marchid = 0x8a45;
 }
 
+static void rv64_andes_ax45mpv_cpu_init(Object *obj)
+{
+    RISCVCPUConfig *cfg = &RISCV_CPU(obj)->cfg;
+    CPURISCVState *env = &RISCV_CPU(obj)->env;
+
+    set_misa(env, MXL_RV64, RVI | RVM | RVA | RVF | RVD | RVC | RVS | RVU | RVV);
+    rv64_andes_common_cpu_init(obj, andes_set_mmsc_cfg_l2c);
+
+    /* Set CPU ID */
+    cfg->marchid = 0x8a45;
+
+    /* Bitmanip */
+    cfg->ext_zba = true;
+    cfg->ext_zbb = true;
+    cfg->ext_zbc = true;
+    cfg->ext_zbs = true;
+
+    /* Vector */
+    cfg->ext_v = true;
+    cfg->vext_spec = g_strdup("v1.0");
+    cfg->vlen  = 1024;
+    cfg->elen  = 64;
+}
+
 static void rv64_andes_ax65_cpu_init(Object *obj)
 {
     RISCVCPUConfig *cfg = &RISCV_CPU(obj)->cfg;
@@ -2110,6 +2134,7 @@ static const TypeInfo riscv_cpu_type_infos[] = {
     DEFINE_CPU(TYPE_RISCV_CPU_ANDES_AX25,       rv64_andes_ax25_cpu_init),
     DEFINE_CPU(TYPE_RISCV_CPU_ANDES_AX27,       rv64_andes_ax27_cpu_init),
     DEFINE_CPU(TYPE_RISCV_CPU_ANDES_AX45,       rv64_andes_ax45_cpu_init),
+    DEFINE_CPU(TYPE_RISCV_CPU_ANDES_AX45MPV,    rv64_andes_ax45mpv_cpu_init),
     DEFINE_CPU(TYPE_RISCV_CPU_ANDES_AX65,       rv64_andes_ax65_cpu_init),
     DEFINE_CPU(TYPE_RISCV_CPU_ANDES_NX25,       rv64_andes_nx25_cpu_init),
     DEFINE_CPU(TYPE_RISCV_CPU_ANDES_NX27V,      rv64_andes_nx27v_cpu_init),
-- 
2.34.1

